### PR TITLE
Add powershell param

### PR DESCRIPTION
**Pull request by mateimicu**
This is not ready! I tested this on one Scenario. I made the pull request to know if the implementation it's okay.
- [x] Tested on a Scenario
- [x] Tested on all Scenarios

### CODE REVIEW

**Comment by alexcoman** [comment]
Please change the commit message for the following commits: https://github.com/cloudbase/cloudbase-init-ci/pull/5/commits/e2681d5602140478cb3d964e8612678c326e340c, https://github.com/cloudbase/cloudbase-init-ci/pull/5/commits/3e6e7258424e8430ceec169109a7f70561ed3172.

File: argus/util.py [hunk 1]
```diff
@@ -242,6 +246,36 @@ def get_certificate():
...
+              "-EncodedCommand -Command {}".format(encoded)
+    return command
+
+COMMAND_MODIFIERS = {
+        POWERSHELL: _get_command_POWRSHELL,
+        POWERSHELL_BYPASS: _get_command_POWRSHELL_BYPASS,
+        }
+
+
+def get_command(command, command_type=None):
```
Comment (alexcoman) [inline argus/util.py:296]: Please add a docstring for it.

File: argus/util.py [hunk 2]
```diff
@@ -242,6 +246,36 @@ def get_certificate():
...
+
+    return command
+
+
+def _get_command_POWRSHELL_BYPASS(command):
+    encoded = base64.b64encode(command.encode("UTF-16LE"))
+    if six.PY3:
+        encoded = encoded.decode()
+
+    command = "powershell -Noninteractive -NoLogo -ExecutionPolicy ByPass" + \
```
Comment (alexcoman) [inline argus/util.py]: I think you should avoid string concatenation.

``` python
command = ("powershell -Noninteractive -NoLogo -ExecutionPolicy ByPass"
           "-EncodedCommand -Command {}").format(encoded)
```

File: argus/util.py [hunk 3]
```diff
@@ -242,6 +246,36 @@ def get_certificate():
     return get_resource("certificate")
 
 
+def _get_command_POWRSHELL(command):
+    encoded = base64.b64encode(command.encode("UTF-16LE"))
```
Comment (alexandrucoman) [inline argus/util.py]: This function contain redundant code.

File: argus/util.py [hunk 4]
```diff
@@ -242,6 +246,36 @@ def get_certificate():
...
+    if six.PY3:
+        encoded = encoded.decode()
+
+    command = "powershell -Noninteractive -NoLogo" + \
+              "-EncodedCommand -Command {}".format(encoded)
+
+    return command
+
+
+def _get_command_POWRSHELL_BYPASS(command):
```
Comment (alexcoman) [inline argus/util.py]: This function contain redundant code.

**Comment by alexcoman** [comment]
You should use commit messages that are relevant. Please change the commit message for the following commits: https://github.com/cloudbase/cloudbase-init-ci/pull/5/commits/765176e1d9448f0efa443a3d20607b551f513a51 https://github.com/cloudbase/cloudbase-init-ci/pull/5/commits/6127f1fec79e893b2c3002f335405568e0759671 https://github.com/cloudbase/cloudbase-init-ci/pull/5/commits/d0df9e43a8d9b2d5c70edf192cb8455700a25095 https://github.com/cloudbase/cloudbase-init-ci/pull/5/commits/41e161113f207c74474329b0feaa58a91e337488 https://github.com/cloudbase/cloudbase-init-ci/pull/5/commits/24ff4ec7e08e9987df21aedf222038cb33619fe7.

File: argus/util.py [hunk 5]
```diff
@@ -242,6 +246,30 @@ def get_certificate():
...
+               " -EncodedCommand {}").format(encoded)
+
+    return command
+
+
+def _get_command_POWRSHELL_SCRIPT(command):
+    return "powershell -File {}".format(command)
+
+COMMAND_MODIFIERS = {
+        POWERSHELL: _get_command_POWRSHELL,
```
Comment (mateimicu) [inline argus/util.py]: Bad indentation.

File: argus/util.py [hunk 6]
```diff
@@ -242,6 +246,30 @@ def get_certificate():
     return get_resource("certificate")
 
 
+def _get_command_POWRSHELL(command):
```
Comment (alexpilotti) [inline argus/util.py]: _get_command_POWERSHELL

File: argus/util.py [hunk 7]
```diff
@@ -242,6 +246,30 @@ def get_certificate():
...
+    if six.PY3:
+        encoded = encoded.decode()
+
+    command = ("powershell -Noninteractive -NoLogo"
+               " -EncodedCommand {}").format(encoded)
+
+    return command
+
+
+def _get_command_POWRSHELL_SCRIPT(command):
```
Comment (alexpilotti) [inline argus/util.py]: ditto (typo)

File: argus/client/windows.py [hunk 8]
```diff
@@ -131,44 +139,46 @@ def copy_file(self, filepath, remote_destination):
...
+    def run_command_verbose(self, cmd, command_type=util.POWERSHELL):
         """Run the given command and log anything it returns.
 
         Do this with retrying support.
 
         :rtype: string
         :returns: stdout
         """
-        stdout, stderr, exit_code = self.run_command_with_retry(cmd)
+        rezult = self.run_command_with_retry(cmd, command_type=command_type)
```
Comment (alexcoman) [inline argus/client/windows.py]: Typo. (result instead of rezult).

File: argus/recipes/cloud/windows.py [hunk 9]
```diff
@@ -86,10 +92,11 @@ def _run_installation(self, service_type):
                  "using the installer %r with service %r.",
                  installer, service_type)
 
-        cmd = ('powershell "C:\\\\installcbinit.ps1 -serviceType {} '
-               '-installer {}"'.format(service_type, installer))
+        cmd = ('C:\\\\installcbinit.ps1 -serviceType {} '
```
Comment (alexcoman) [inline argus/recipes/cloud/windows.py]: You can use

``` python
cmd = r'C:\\installcbinit.ps1' # ...
```

File: argus/util.py (hunk 6 above)
Comment (alexcoman) [inline argus/util.py]: Function names should be lowercase, with words separated by underscores as necessary to improve readability.

File: argus/util.py [hunk 10]
```diff
@@ -242,6 +252,53 @@ def get_certificate():
...
+    if six.PY3:
+        encoded = encoded.decode()
+
+    command = ("powershell -Noninteractive -NoLogo"
+               " -EncodedCommand {}").format(encoded)
+
+    return command
+
+
+def _get_command_POWERSHELL_SCRIPT(command):
```
Comment (alexcoman) [inline argus/util.py]: ditto

File: argus/introspection/cloud/windows.py [hunk 11]
```diff
@@ -67,7 +67,8 @@ def _get_ntp_peers(output):
 
 
 def escape_path(path):
-    """Escape the spaces in the given path in order to work with Powershell properly."""
+    """Escape the spaces in the given path in order to work
+       with Powershell properly."""
```
Comment (alexcoman) [inline argus/introspection/cloud/windows.py]: Remove the extra newline from the docstring.

File: argus/recipes/cloud/windows.py [hunk 12]
```diff
@@ -41,30 +41,36 @@ class CloudbaseinitRecipe(base.BaseCloudbaseinitRecipe):
...
 
-        wait_cmd = ('powershell "(Get-WmiObject Win32_Account | '
-                    'where -Property Name -contains {0}).Name"'
-                    .format(self._conf.openstack.image_username))
-        self._execute_until_condition(
-            wait_cmd,
-            lambda stdout: stdout.strip() == self._conf.openstack.image_username,
-            count=COUNT, delay=DELAY)
+        username = self._conf.openstack.image_username
+
```
Comment (alexcoman) [inline argus/recipes/cloud/windows.py]: The newline is not required.

File: argus/recipes/cloud/windows.py [hunk 13]
```diff
@@ -226,24 +238,23 @@ def pre_sysprep(self):
...
                "{}/windows/patch_shell.ps1 -outfile "
                "C:\\patch_shell.ps1"
                .format(self._conf.argus.resources))
-        self._execute(cmd)
+        self._execute(cmd, command_type=util.POWERSHELL)
 
-        escaped = introspection.escape_path(cbinit)
-        self._execute('powershell "C:\\\\patch_shell.ps1 \"{}\""'
-                      .format(escaped))
+        self._execute('C:\\\\patch_shell.ps1 \"{}\"'
```
Comment (alexcoman) [inline argus/recipes/cloud/windows.py]:
``` python
r'C:\\' # ...
```

File: argus/tests/cloud/windows/test_smoke.py [hunk 14]
```diff
@@ -79,15 +83,17 @@ def test_https_winrm_configured(self):
...
+                'echo 1', command_type=util.POWERSHELL)
         self.assertEqual('1', stdout.strip())
 
     @test_util.skip_unless_dnsmasq_configured
     def test_w32time_triggers(self):
         # Test that w32time has network availability triggers, not
         # domain joined triggers
         if self._introspection.get_instance_os_version() > (6, 0):
-            start_trigger, _ = self._introspection.get_service_triggers('w32time')
+            time = "w32time"
```
Comment (alexcoman) [inline argus/tests/cloud/windows/test_smoke.py]: Is not required.

File: argus/introspection/cloud/windows.py [hunk 15]
```diff
@@ -182,8 +186,9 @@ def get_cbinit_key(execute_function):
            "Cloudbase-init")
     key_x64 = ("HKLM:SOFTWARE\\Wow6432Node\\Cloudbase` Solutions\\"
                "Cloudbase-init")
-    cmd = 'powershell "Test-Path {}"'.format(key)
-    if execute_function(cmd).strip().lower() == "true":
+    cmd = 'Test-Path {}'.format(key)
+    rezult = execute_function(cmd, command_type=util.POWERSHELL)
```
Comment (stefan-caraiman) [inline argus/introspection/cloud/windows.py]: Typo (should be result)

File: argus/recipes/cloud/windows.py [hunk 16]
```diff
@@ -41,30 +41,36 @@ class CloudbaseinitRecipe(base.BaseCloudbaseinitRecipe):
...
+        wait_cmd = ('(Get-WmiObject Win32_Account | '
+                    'where -Property Name -contains {0}).Name'
+                    .format(username))
+        try:
+            self._execute_until_condition(
+                wait_cmd,
+                lambda stdout: stdout.strip() == username,
+                count=COUNT, delay=DELAY, command_type=util.POWERSHELL)
+        except exceptions.ArgusCLIError as err:
+            LOG.debug(("We got and ArgusCLIError and it is ignored"
```
Comment (stefan-caraiman) [inline argus/recipes/cloud/windows.py]: Typo (an ArgusCLIError)

File: argus/tests/cloud/windows/test_smoke.py [hunk 17]
```diff
@@ -158,15 +165,27 @@ class TestNextLogonPassword(base.BaseTestCase):
...
             self._conf.openstack.image_password)
+        LOG = util.get_logger()
+        try:
+            remote_client.run_command_until_condition(
+                wait_cmd,
+                lambda out: out.strip() == 'Stopped',
+                retry_count=util.RETRY_COUNT, delay=util.RETRY_DELAY,
+                command_type=util.POWERSHELL)
+        except exceptions.ArgusCLIError as err:
+            LOG.debug(("We got and ArgusCLIError and it is ignored"
```
Comment (stefan-caraiman) [inline argus/tests/cloud/windows/test_smoke.py]: Ditto

File: argus/introspection/cloud/windows.py [hunk 18]
```diff
@@ -192,38 +196,44 @@ class InstanceIntrospection(base.CloudInstanceIntrospection):
...
     def username_exists(self, username):
-        cmd = ('powershell "Get-WmiObject Win32_Account | '
-               'where -Property Name -contains {0}"'
+        cmd = ('Get-WmiObject Win32_Account | '
+               'where -Property Name -contains {0}'
                .format(username))
 
-        stdout = self.remote_client.run_command_verbose(cmd)
+        stdout = self.remote_client.run_command_verbose(
+                cmd, command_type=util.POWERSHELL)
```
Comment (alexcoman) [inline argus/introspection/cloud/windows.py]:
```
C:210, 0: Wrong hanging indentation.
                cmd, command_type=util.POWERSHELL)
            |   ^ (bad-continuation)
```

File: argus/introspection/cloud/windows.py [hunk 19]
```diff
@@ -192,38 +196,44 @@ class InstanceIntrospection(base.CloudInstanceIntrospection):
...
         return ntpath.join(
             homedir, self._conf.cloudbaseinit.created_user,
             ".ssh", "authorized_keys")
 
     def get_instance_file_content(self, filepath):
-        cmd = 'powershell "cat %s"' % filepath
-        return self.remote_client.run_command_verbose(cmd)
+        cmd = 'cat %s' % filepath
+        return self.remote_client.run_command_verbose(
+                cmd, command_type=util.POWERSHELL)
```
Comment (alexcoman) [inline argus/introspection/cloud/windows.py]: ditto

File: argus/introspection/cloud/windows.py [hunk 20]
```diff
@@ -192,38 +196,44 @@ class InstanceIntrospection(base.CloudInstanceIntrospection):
...
 
     def get_userdata_executed_plugins(self):
-        cmd = 'powershell "(Get-ChildItem -Path  C:\\ *.txt).Count'
-        stdout = self.remote_client.run_command_verbose(cmd)
+        cmd = '"(Get-ChildItem -Path  C:\\ *.txt).Count'
+        stdout = self.remote_client.run_command_verbose(
+                cmd, command_type=util.POWERSHELL)
```
Comment (alexcoman) [inline argus/introspection/cloud/windows.py]: ditto

File: argus/introspection/cloud/windows.py [hunk 21]
```diff
@@ -245,7 +255,8 @@ def _parse_netsh_output(output):
 
     def get_instance_mtu(self):
         cmd = 'netsh interface ipv4 show subinterfaces level=verbose'
-        stdout = self.remote_client.run_command_verbose(cmd)
+        stdout = self.remote_client.run_command_verbose(
+                cmd, command_type=util.CMD)
```
Comment (alexcoman) [inline argus/introspection/cloud/windows.py]: ditto

File: argus/introspection/cloud/windows.py [hunk 22]
```diff
@@ -254,20 +265,22 @@ def get_cloudbaseinit_traceback(self):
...
         return stdout.strip() == 'True'
 
     def instance_exe_script_executed(self):
         return self._file_exist("C:\\Scripts\\exe.output")
 
     def get_group_members(self, group):
         cmd = "net localgroup {}".format(group)
-        std_out = self.remote_client.run_command_verbose(cmd)
+        std_out = self.remote_client.run_command_verbose(
+                    cmd, command_type=util.CMD)
```
Comment (alexcoman) [inline argus/introspection/cloud/windows.py]: ditto

**Comment by AlexandruTudose** [comment]
LGTM

File: argus/recipes/cloud/windows.py [hunk 23]
```diff
@@ -41,30 +41,35 @@ class CloudbaseinitRecipe(base.BaseCloudbaseinitRecipe):
...
+                    'where -Property Name -contains {0}).Name'
+                    .format(username))
+        try:
+            self._execute_until_condition(
+                wait_cmd,
+                lambda stdout: stdout.strip() == username,
+                count=COUNT, delay=DELAY, command_type=util.POWERSHELL)
+        except exceptions.ArgusCLIError as err:
+            LOG.debug(("We got an ArgusCLIError and it is ignored"
+                      " with the output : {}").format(err.message))
```
Comment (alexcoman) [inline argus/recipes/cloud/windows.py]:
```
C: 55, 0: Wrong continued indentation.
                      " with the output : {}").format(err.message))
                      ^| (bad-continuation)
```

File: argus/recipes/cloud/windows.py [hunk 24]
```diff
@@ -265,22 +275,33 @@ def _wait_cbinit_finalization(self, searched_paths=None):
...
+        try:
+            self._execute_until_condition(
+                wait_cmd,
+                lambda out: out.strip() == 'Stopped',
+                count=COUNT, delay=DELAY, command_type=util.POWERSHELL)
+        except exceptions.ArgusCLIError as err:
+            # This error ocures when you execute the first command when you boot 
+            # the machine
+            LOG.debug(("We got and ArgusCLIError and it is ignored"
+                      " with the output : {}").format(err.message))
```
Comment (alexcoman) [inline argus/recipes/cloud/windows.py]: Use `%` formatting in logging functions but pass the `%` parameters as arguments (logging-format-interpolation).

**Pull request merged**